python/cudf/cudf/tests/conftest.py

```diff
@@ -15,6 +20,106 @@ def datadir():
     return pathlib.Path(__file__).parent / "data"
 
 
+@pytest.fixture(
+    params=itertools.product([0, 2, None], [0.3, None]),
+    ids=lambda arg: f"n={arg[0]}-frac={arg[1]}",
+)
+def sample_n_frac(request):
+    """
+    Specific to `test_sample*` tests.
+    """
+    n, frac = request.param
+    if n is not None and frac is not None:
+        pytest.skip("Cannot specify both n and frac.")
+    return n, frac
+
+
+def _random_state_tuple(seed_or_state_ctor):
+    def shape_checker(expected, got):
+        assert expected.shape == got.shape
+
+    def exact_checker(expected, got):
+        assert_eq(expected, got)
+
+    if seed_or_state_ctor is None:
+        return None, None, shape_checker
+    elif isinstance(seed_or_state_ctor, int):
+        return seed_or_state_ctor, seed_or_state_ctor, shape_checker
+    elif seed_or_state_ctor == np.random.RandomState:
+        return seed_or_state_ctor(42), seed_or_state_ctor(42), exact_checker
+    elif seed_or_state_ctor == cp.random.RandomState:
+        return np.random.RandomState(42), seed_or_state_ctor(42), shape_checker
+
+    assert False, "No random state is specialized beyond the above types."
+
+
+@pytest.fixture(
+    params=[None, 42, np.random.RandomState],
+    ids=["None", "IntSeed", "NumpyRandomState"],
+)
+def random_state_tuple_axis_1(request):
+    """
+    Specific to `test_sample*_axis_1` tests.
+    A pytest fixture of valid `random_state` parameter pairs for pandas
+    and cudf. Valid parameter combinations, and what to check for each pair
+    are listed below:
+
+    pandas:   None,   seed(int),  np.random.RandomState
+    cudf:     None,   seed(int),  np.random.RandomState
+    ------
+    check:    shape,  shape,      exact result
+
+    Each column above stands for one valid parameter combination and check.
+    """
+
+    return _random_state_tuple(request.param)
+
+
+@pytest.fixture(
+    params=[None, 42, np.random.RandomState, cp.random.RandomState],
+    ids=["None", "IntSeed", "NumpyRandomState", "CupyRandomState"],
+)
+def random_state_tuple_axis_0(request):
+    """
+    Specific to `test_sample*_axis_0` tests.
+    A pytest fixture of valid `random_state` parameter pairs for pandas
+    and cudf. Valid parameter combinations, and what to check for each pair
+    are listed below:
+
+    pandas:   None,   seed(int),  np.random.RandomState,  np.random.RandomState
+    cudf:     None,   seed(int),  np.random.RandomState,  cp.random.RandomState
+    ------
+    check:    shape,  shape,      exact result,           shape
+
+    Each column above stands for one valid parameter combination and check.
+    """
+
+    return _random_state_tuple(request.param)
+
+
+@pytest.fixture(params=[None, "builtin_list", "ndarray"])
+def make_weights_axis_0(request, random_state_tuple_axis_0):
+    """Specific to `test_sample*_axis_0` tests.
+    Only testing weights array that matches type with random state.
+    """
+    _, gd_random_state, _ = random_state_tuple_axis_0
```

Comment:
Can we rewrite this fixture to not use the `random_state_tuple_axis_0`? It looks like two of the conditional branches below never use this output anyway, so we generate too many cases, and we also only use a subset of the possible values when checking the type of the `gd_random_state` in `wrapped`.

Reply:
Yes, I should be able to rewrite this to return a factory that accepts an extra argument for distinction.

> so we generate too many cases,

This fixture request does not generate more test cases. The number of test cases is determined by the product of all parametrization of all fixtures requested by the test case. Here, for each of the test case it is only requesting the instantiation of the random state tuple. Here's a small demo:

```python
@pytest.fixture(params=[1, 2])
def fixture1(request):
    return request.param

@pytest.fixture(params=['a', 'b'])
def fixture2(request, fixture1):
    return fixture1


def test_f(fixture1, fixture2):
    pass
```

```bash
(rapids) rapids@compose:~/scratch$ pytest --collect-only pytest_play/fixture/parametrized_fixture.py 
========================================================== test session starts ===========================================================
platform linux -- Python 3.8.12, pytest-7.0.1, pluggy-1.0.0
benchmark: 3.4.1 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
rootdir: /home/yhw/dev/rapids/scratch
plugins: forked-1.4.0, xdist-2.5.0, pudb-0.7.0, benchmark-3.4.1, hypothesis-6.36.2, repeat-0.9.1
collected 4 items                                                                                                                        

<Module pytest_play/fixture/parametrized_fixture.py>
  <Function test_f[1-a]>
  <Function test_f[1-b]>
  <Function test_f[2-a]>
  <Function test_f[2-b]>
```

Reply:
Yes, but in this case some of those parametrizations are redundant. For example, you will generate four different tests where `request.param` is `None` because `random_state_tuple_axis_0` has four possible values, but all of those tests will do exactly the same thing. The same thing happens when `request.param == "builtin-list"`.

Reply:
I would agree that part of the parametrization of random state tuple for `make_weights` fixture is redundant. If we are to be very conservative about introducing unnecessary context to another method, the state of PR is changed to reflect that.

For pure academic debates, it can also be argued that requesting random state fixture explicitly shows the dependency of `make_weights` fixture has on the random state fixture, while the state of PR isn't as clear.

Reply:
I think there exists a path to make both world happy - one would require parameterize pandas random state, cudf random state and checker independently into their own fixture and express their dependency with a forth fixture. Just a bit more work to figure out the constraints there.